feat(ui-licenseobligation): License obligation changes

smrutis1 opened the pull request:
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.

Please refer the mentioned issue.

> * Which issue is this pull request belonging to and how is it solving it? (#1050 )
> * Did you add or update any new dependencies that are required for your change? - No

Issue:  
closes #1050 

### Suggest Reviewer
> You can suggest reviewers here with an @mention.

### How To Test?
> How should these changes be tested by the reviewer? 

* Need to test the editing the license and adding obligation to it.

> Have you implemented any additional tests? - No

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

Signed-off-by: Smruti Prakash Sahoo <smruti.sahoo@siemens.com>

JaideepPalit reviewed (changes requested):
Reviewed the PR

JaideepPalit commented:
Tested the PR. Observed some issues.

1. Update Button is coming in bottom of the table. Cancel button not there in Linked obligation Page
**Issue**
![image](https://user-images.githubusercontent.com/56534710/100847912-6cd44c00-34a6-11eb-9255-57c1a52eb835.png)
**Expected**
![image](https://user-images.githubusercontent.com/56534710/100848065-a311cb80-34a6-11eb-8be9-130331847cbd.png)

2. Update not working as expected. Update Fields in both the tab`License and Linked Obligation`  -> Goto linked obligation -> click on update.
Issue 1- License Tab changes not updated
Issue 2 - Adds all obligation to license instead of selected one.
Expected behavior - There should be one Update button to update entire License/

3. Linked Obligation Table in Edit Page
    - Add Obligation type in table. 
    - sort not working for all relevant fields . 
![image](https://user-images.githubusercontent.com/56534710/100848436-1e737d00-34a7-11eb-8155-ed561e4844ac.png)

4. Linked Obligation table in Details Page
![image](https://user-images.githubusercontent.com/56534710/100848823-ab1e3b00-34a7-11eb-80ee-6d237fcc779f.png)
Is it possible to add/update `Applies to Development`, `Applies to Distribution` and `Further Properties`?
Kindly remove fields from code which are not used anymore. Maintain same table in Edit and Details Page

5. Maintain Similar table Structure and Functionality for Details/Edit and Add Obligation after excluding irrelevant columns
![image](https://user-images.githubusercontent.com/56534710/100850049-406dff00-34a9-11eb-921d-2169eeb9c29d.png)

6. Select Obligation Dialog Box
    - Sorting Searching is Disabled
    - Add Obligation Type
    - Select all Obligation checkbox missing
    - Break work for very long obligation inorder to maintain proper structure of Dialog box
![image](https://user-images.githubusercontent.com/56534710/100849267-5202d700-34a8-11eb-8b04-b07672f87e5c.png)

7. Able to add same obligation multiple times
![image](https://user-images.githubusercontent.com/56534710/100849321-68109780-34a8-11eb-93ac-6f7c696f8edf.png)

8. Delete Button not working

smrutis1 commented on scripts/migrations/042_remove_validForProject_from_Obligation.py:
```diff
@@ -0,0 +1,83 @@
+    print 'Removing field name '+fieldToBeRemoved
+    log['Updated obligation fields '+fieldToBeRemoved] = []
+    for entity in qryResult:
+        del entity[''+fieldToBeRemoved+'']
```
Changed it

smrutis1 commented on libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java:
```diff
@@ -612,7 +612,7 @@ public static Integer parseStringToNumber(String input) {
 
         Map<String, ObligationStatusInfo> mapOfObligations = obligations.stream().filter(Objects::nonNull)
-                .filter(o -> isValidForProjectOptional || o.isValidForProject())
+                .filter(o -> isValidForProjectOptional)
```
removed

smrutis1 commented on frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/includes/detailAddTodo.jspf:
```diff
@@ -153,7 +201,51 @@
+                    oblIds.push(oblId);
+                    $OlTitle = $('td:eq(1)', this);
+                    $OlText = $('td:eq(2)', this);
+                    row = "<tr><td>" + $OlTitle.text() + "</td><td>" + $OlText.text() + "</td><td class='content-middle'><svg class='action lexicon-icon' data-row-id='oblLinkRow${ob.id}' data-obligation-title='${ob.title}'><title><liferay-ui:message key='delete' /></title><use href='/o/org.eclipse.sw360.liferay-theme/images/clay/icons.svg#trash'/></svg></td></tr>"
```
Done

smrutis1 commented on frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/includes/detailAddTodo.jspf:
```diff
@@ -19,79 +19,127 @@
+			</div>
+		</div>
+	</div>
+</div>
```
Have used `$dialog = dialog.confirm()`

smrutis1 commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java:
```diff
@@ -141,8 +142,27 @@ private void prepareEditView(RenderRequest request, RenderResponse response) {
+
+            if (licenseOblIds != null) {
+                obls = CommonUtils.nullToEmptyList(client.getObligationsByIds(Lists.newArrayList(licenseOblIds)));
+            }
```
Done

smrutis1 commented on frontend/sw360-portlet/src/main/resources/META-INF/resources/html/licenses/edit.jsp:
```diff
@@ -50,7 +50,23 @@
+                            </div>
+                            <div id="tab-AddLicense" class="tab-pane <core_rt:if test="${selectedTab == 'tab-AddLicense'}">active show</core_rt:if>">
+                            <div class="row portlet-toolbar">
             <div class="col-auto">
```
Done

smrutis1 commented:
Thanks for review and suggestions. Have addressed most of the issues. Kindly have a look and share the feedback(if any).

Regarding point 4.  Have removed `Applies to Development`, `Applies to Distribution`  from the UI, But have kept the code for now in the backend as need to test the license import functionality also after removing these.

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/licenses/LicensesPortlet.java:
```diff
@@ -317,6 +359,9 @@ private License updateLicenseFromRequest(License license, ActionRequest request)
         license.setGPLv3Compat(gpl3compatibility);
         license.setChecked(checked);
+        String obligationIds = request.getParameter("obligations");
+        List<String> oblIds = obligationIds == null || obligationIds.equals("") ? Lists.newArrayList() : Arrays.asList(obligationIds.split(","));
```
Is it possible to use `CommonUtils.isNotNullEmptyOrWhitespace()` same for line 457

smrutis1 replied:
Have used

JaideepPalit commented on libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java:
```diff
@@ -612,7 +612,6 @@ public static Integer parseStringToNumber(String input) {
         obligationAlreadyPresent.entrySet().stream().forEach(e -> obligationStatusMap.remove(e.getKey()));
 
         Map<String, ObligationStatusInfo> mapOfObligations = obligations.stream().filter(Objects::nonNull)
-                .filter(o -> isValidForProjectOptional || o.isValidForProject())
```
Should also be removed from method signature.

smrutis1 replied:
Removed it

JaideepPalit reviewed (changes requested):
Code looks fine to me , expect below two comments.

smrutis1 commented:
Thanks for comments. Have made the changes

JaideepPalit reviewed (approved):
Code looks good to me. Needs retest.

mcjaeger commented:
restested - adding obligation from creation before from license admin portlets works. obligations can be added to project

mcjaeger commented:
@JaideepPalit please consider merging after (if) travis wen through
